pyxpp/ast_builder/ast_builder.py

```diff
@@ -0,0 +1,633 @@
+from typing import Union, Optional, List
+from enum import Enum
+from pyxpp.recognizer.token import Token, TokenType
+from .ast_node import ASTNode, Program, Comment, Variable, FunctionCall, Operation, FunctionDefinition, Return, Class, IfStatement, WhileLoop, ForLoop, NodeWithChildren, FromImportStatement, BreakStatement, ContinueStatement, ImportStatement
+
+class ASTBuilder:
+    def __init__(self):
+        """
+        Initializes the ASTBuilder instance, setting up the token list, root node,
+        node order, and the initial tabulation level.
+
+        Attributes:
+            tokens (list): A list to hold the tokens for building the AST.
+            root (ASTNode | None): The root node of the AST, initially set to None.
+            node_order (list): A stack to manage the current traversal state of the AST.
+            last_tabulation (int): Tracks the level of indentation in the code.
+        """
+        self.tokens = []
+        self.root = None
+        self.node_order = []
+        self.last_tabulation = 0
+
+    def process_token(self, token: Token) -> None:
+        """
+        Adds a token to the internal list of tokens for the AST building process.
+
+        Args:
+            token (Token): The token to be processed and added.
+        """
+        self.tokens.append(token)
+
+    def build(self) -> ASTNode:
+        """
+        Constructs the AST from the provided tokens and sets the root node.
+
+        Returns:
+            ASTNode: The root node of the constructed AST.
+        """
+        self.root = Program()
+        self.node_order = [self.root]
```

Comment:
`self.node_order.append(self.root)`
 There is no need to re-assign here.

Reply:
done